cmd/restic/cmd_web.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/restic/restic"
+	"github.com/restic/restic/backend"
+	"github.com/restic/restic/cmd/restic/web"
+	"github.com/restic/restic/repository"
+)
+
+type CmdWeb struct {
+	global *GlobalOptions
+}
+
+func init() {
+	_, err := parser.AddCommand("server",
+		"serve repository in a web interface",
+		"The web command serves the repository in a web interface",
+		&CmdWeb{global: &globalOpts})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (cmd CmdWeb) Usage() string {
+	return "PORT"
+}
+
+func (cmd CmdWeb) Execute(args []string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("port not specified, usage:%s", cmd.Usage())
+	}
+
+	repo, err := cmd.global.OpenRepository()
+	if err != nil {
+		return err
+	}
+	repo.LoadIndex()
+	httpRepo := &httpRepo{
+		repo: repo,
+		objects: map[string]http.File{
+			"": &httpRoot{repo: repo},
+		},
+	}
+
+	portStr := args[0]
+	if port, err := strconv.Atoi(portStr); err != nil || port > 65535 {
+		return fmt.Errorf("%s is not a valid port number", port)
+	}
+
+	cmd.global.Printf("Now serving on %s\n", portStr)
+	http.ListenAndServe(":"+portStr, http.FileServer(httpRepo))
+	return nil
+}
+
+type httpRepo struct {
+	repo *repository.Repository
+
+	// Object path -> object
+	objects map[string]http.File
+}
+
+func (hr httpRepo) Open(name string) (http.File, error) {
+	var object http.File
+	var found bool
+
+	parts := strings.Split(name, "/")
+	for _, part := range parts {
+		object, found = hr.objects[part]
+		if !found {
+			return nil, fmt.Errorf("%s not found", name)
+		}
+		children, err := object.(childrener).loadChildren()
+		if err != nil {
+			return nil, err
+		}
+		for _, child := range children {
+			if _, ok := hr.objects[child.Name()]; !ok {
+				hr.objects[child.Name()] = child
```

Comment:
High risk of collision here: we only use the node's name, we should use its full path